### PR TITLE
Improve typing and change StyleLayer query call interface

### LINKED ISSUE #5255: Globe Projection Bug: Circle Layers have smaller hover/click area than visually represented
**maplibre-gl-js version**: 5.0.0-pre10

**browser**: Reproduced in Chrome and Firefox on Ubuntu 24.10

### Steps to Trigger Behavior

 1. Create a Circle Layer with Icons OR actual circles
 2. When using in mercator mode the hover/click events match the rendered symbol's bounds
 3. Turning on globe projects seems to only allow a very small point within the rendered circle or icon that hovers or clicks

### Link to Demonstration

https://codepen.io/ingalls-the-bold/pen/jENmLGe

### Expected Behavior

MouseEnter/Leave & Click covers entire rendered feature

### Actual Behavior

MouseEnter/Lease & Click is much smaller than actual rendered feature in the new Globe projection

### PR DESCRIPTION
**Pull request by HarelM**
## Launch Checklist

This is a pure refactoring step to add more types to the query code flows.
It also add the globe projection matrix to be passed down to the style layers that needs it, but it doesn't solve any problems with exiting code (mainly circle layer).

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by HarelM** [comment]
I've done a very simple change in the position matrix that is passed to the `queryIntersectsFeature` of every layer.
This seems to still pass the tests.
I now see that there are missing tests for query stuff related to globe (`test/integration/query`).
I'm still not sure how to "see" what I'm writing in terms how to present on the map the intersection perimeter, so I'm still in very early stages of this fix.
Any help would be greatly appreciated.
Cc: @ibesora @kubapelc

**Comment by HarelM** [comment]
I was able to leverage the following method from the transform class:
`projectTileCoordinates` to project tile coordinates to screen coordinates, but I'm not sure it a great solution.
image

In order to use it I need to pass down the tileID, which is not a big deal since the tile is the caller in this case, and it know its own ID.
I can check it in if this would help anyone to better understand what I did.
I think it is similar to @ibesora fix here:  #5267, and it break the same way as this PR when adding `circle-translate`...

**Comment by HarelM** [comment]
Ignore my last comment, it breaks mercator related circle stuff...

**Comment by HarelM** [comment]
This is now ready for review, I'll add comments where I changed something that is not only types.

File: src/source/query_features.ts [hunk 1]
```diff
@@ -122,19 +132,7 @@ export function queryRenderedFeatures(
...
-            const state = sourceCache.getFeatureState(feature.layer['source-layer'], feature.id);
-            feature.source = feature.layer.source;
-            if (feature.layer['source-layer']) {
-                feature.sourceLayer = feature.layer['source-layer'];
-            }
-            feature.state = state;
-        });
-    }
-    return result;
+    return convertFeaturesToMapFeatures(result, sourceCache);
```
Comment (HarelM) [inline src/source/query_features.ts:135]: I've moved this code to the end of the file to be somewhat reused.

File: src/style/style_layer.ts [hunk 2]
```diff
@@ -56,16 +94,7 @@ export abstract class StyleLayer extends Evented {
...
-        queryGeometry: Array<Point>,
-        feature: VectorTileFeature,
-        featureState: FeatureState,
-        geometry: Array<Array<Point>>,
-        zoom: number,
-        transform: IReadonlyTransform,
-        pixelsToTileUnits: number,
-        pixelPosMatrix: mat4
-    ): boolean | number;
+    queryIntersectsFeature?(params: QueryIntersectsFeatureParams): boolean | number;
```
Comment (HarelM) [inline src/style/style_layer.ts:97]: This an interface change, but it only moves the method parameters to a "type".

File: src/source/query_features.ts [hunk 3]
```diff
@@ -55,18 +62,21 @@ export type QuerySourceFeatureOptions = {
...
+function getPixelPosMatrix(transform: IReadonlyTransform, tileID: OverscaledTileID) {
     const t = mat4.create();
     mat4.translate(t, t, [1, 1, 0]);
     mat4.scale(t, t, [transform.width * 0.5, transform.height * 0.5, 1]);
-    if (transform.calculatePosMatrix) { // Globe: TODO: remove this hack once queryRendererFeatures supports globe properly
-        return mat4.multiply(t, t, transform.calculatePosMatrix(tileID.toUnwrapped()));
-    } else {
-        return t;
-    }
+    const projectionData = transform.getProjectionData({overscaledTileID: tileID});
```
Comment (HarelM) [inline src/source/query_features.ts:78]: This is the main change in this code which uses the projection data from the transform and adds typings to this method.
As far as I can see this looks OK.
I have removed the `toUnwrapped()` which I'm not sure how it affects the mercator stuff, but the tests do not seem to fail, so I'm not sure if it was needed or not...

Comment (ibesora) [reply src/source/query_features.ts:78]: I think `toUnwrapped` still needs to be there. We are using `tileID.key` as the key into a cache of matrices. Those matrices are the same wether it's wrapped or not but removing `toUnwrapped` will make the cache possibly double in size

**Review by ibesora — APPROVED** [review]
Added a minor nit, otherwise this is a nice code improvement!

**Comment by HarelM** [comment]
The problem is that get projection data doesn't support passing unwrapped. I tried changing that but it didn't go well. 
This part needs more work anyway, so I'll revert that and push only the type changes and minor refactoring ATM...

File: src/source/query_features.ts (hunk 3 above)
Comment (HarelM) [reply src/source/query_features.ts]: I've reverted this.

**Pull request merged**